README.md
## @caripizza/ui
A design system and component library made with React, Typescript, and Tailwind CSS.

[Storybook-main](https://main--63376424914401d9d4d73e3b.chromatic.com/)\
[📦 npm](https://www.npmjs.com/package/@caripizza/ui) / [yarn](https://yarnpkg.com/package/@caripizza/ui)

## Get started

Install the UI library:
```bash
yarn add @caripizza/ui
# and peerDependencies in package.json
```

Add the stylesheet import to your entry-point file:
```tsx
import '@caripizza/ui/dist/output.css';

// or via html/CDN:
<link rel="stylesheet" type="text/css" href="https://unpkg.com/@caripizza/ui@latest/dist/output.css"/>
```
---
## Features

### Typography
`@caripizza/ui` uses the Google [Arimo](https://fonts.google.com/share?selection.family=Arimo:wght@400;500;600) font family, and exposes the following [CSS Vars for fonts](./src/stories/atoms/Typography.stories.mdx):

| CSS Var | Value |
|---|---|
| `--font-sans` | `'Arimo', ui-sans-serif, system-ui, 'Helvetica Neue', sans-serif` |
| `--font-mono` | `'Menlo', ui-monospace, SFMono-Regular, 'Courier New', monospace` |
| `--font-weight-regular` | 400 |
| `--font-weight-medium` | 500 |
| `--font-weight-semibold` | 600 |
| `--font-size-xs` | `0.75rem (12px)` |
| `--font-size-sm` | `0.875rem (14px)` |
| `--font-size-base` | `1rem (16px)` |
| `--font-size-lg` | `1.125rem (18px)` |
| `--font-size-xl` | `1.25rem (20px)` |
| `--font-size-2xl` | `1.5rem (24px)` |
| `--font-size-3xl` | `1.875rem (30px)` |
| `--font-size-4xl` | `2.25rem (36px)` |
| `--font-size-5xl` | `3rem (48px)` |
| `--font-size-6xl` | `3.75rem (60px)` |

### Colors
The stylesheet also includes CSS Vars for colors:
| CSS Var | Description |
|---|---|
| `--color-turquoise` | primary button background colors |
| `--color-gold` | focus state outline color, secondary button |
| `--color-hot-pink` | error button color |
| `--black-100` | light mode font color and dark mode background color |
| `--white-100` | light mode background color and dark mode font color |
| `--text-primary` | black if light mode, white if dark mode |
| `--backg-primary` | white if light mode, black if dark mode |

<br/>

Theme colors are also offered as hex codes:
```jsx
import { colors } from '@caripizza/ui';
// ...
<div
  style={{
    background: `linear-gradient(45deg, ${colors.dark.turquoise[100]}, ${colors.dark.gold[50]})`
  }}
>
```

### Dark mode
All components in the library include dark and light theme styles. To toggle dark mode, append the `dark` class name to an ancestor in the DOM and components will update their UI's accordingly:
```html
<body class="bg-backg-primary text-primary">
  <h1>I render with dark background if system theme is dark</h1>
  <script type="text/javascript">
    window.onload = () => {
      if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.className += " dark";
      }
    }
  </script>
</body>
```

The library also includes utils for detecting a user's system theme:
```tsx
import { hasDarkSystemTheme } from '@caripizza/ui';

if (hasDarkSystemTheme()) {
  toggleDarkMode(true); // etc...
}
```
See the [Dark mode stories](./src/stories/theme) for detailed examples.

----
## Notes

### Support matrix
| Browser | Version | OS
|---|---|---|
| Chrome | Latest version | Mac OS X 12.6 |
| Safari | Latest version | Mac OS X 12.6 |
| Firefox | Latest version | Mac OS X 12.6 |
| Edge | Latest version | Mac OS X 12.6 |

| Framework | Version |
|---|---|
| React | v17, v18 |
| NextJS | v12 |

----
## Contributing
- [CONTRIBUTING](./CONTRIBUTING.md)

----

## Resources
- [WIKI](https://github.com/caripizza/ui/wiki/Resources)
- [LICENSE](./LICENSE.txt)
- [CHANGELOG](./CHANGELOG.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### [0.1.7] - 11-11-2022
- adds Emoji to molecules/ and unicode emoji library/utils 🤖
- adds typography to atoms/, theme
- adds Cypress e2e and component tests, docs, dependencies

### [0.1.6] - 10-20-2022
- adds CommonJS output to dist (prevent need for additional tooling in Nextjs and jest in order to support the ESM output)
- cleans up Storybook docs, tailwind config
- moves `react` and `react-dom` production dependencies to be peer dependencies only

### [0.1.5] - 10-18-2022
- adds `hot-pink` color to theme (#11), tweaks dark mode `turquoise` color
- adds `disabled`, `selected`, `outlined`, and `error` props + styles to Button
- adds semi-transparent `white-alpha` and `black-alpha` colors for disabled states

### [0.1.4] - 10-12-2022
- adds global CSS vars for font sizing and weights
- refines font families/weights/sizes in tailwind config, excludes serif fonts
- cleans up docs

### [0.1.3] - 10-9-2022
- adds lint-staged and Storybook tooling
- refactors colors
- adds `variant` and `loading` props and secondary button variant
- updates light/dark themes

### [0.1.2] - 10-2-2022
- Adds support for Tailwind CSS (#4)
- Adds `hasDarkSystemTheme` to utils/
- Adds colors to atoms/

### [0.1.1] - 10-1-2022
- Install additional tooling, config files

### [0.1.0] - 10-1-2022
- Initial release of library scaffolding, Button

tailwind.config.js
/* eslint-disable @typescript-eslint/no-var-requires */
const plugin = require('tailwindcss/plugin');
const colors = require('./colors');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ['./src/**/*.{tsx,ts}'],
  darkMode: 'class',
  theme: {
    fontWeight: {
      regular: 400,
      medium: 500,
      semibold: 600,
    },
    fontFamily: {
      sans: ['var(--font-sans)'],
      mono: ['var(--font-mono)'],
    },
    fontSize: {
      xs: ['var(--font-size-xs)', { lineHeight: '1rem' }],
      sm: ['var(--font-size-sm)', { lineHeight: '1.25rem' }],
      base: ['var(--font-size-base)', { lineHeight: '1.5rem' }],
      lg: ['var(--font-size-lg)', { lineHeight: '1.75rem' }],
      xl: ['var(--font-size-xl)', { lineHeight: '1.75rem' }],
      '2xl': ['var(--font-size-2xl)', { lineHeight: '2rem' }],
      '3xl': ['var(--font-size-3xl)', { lineHeight: '2.25rem' }],
      '4xl': ['var(--font-size-4xl)', { lineHeight: '2.5rem' }],
      '5xl': ['var(--font-size-5xl)', { lineHeight: '1' }],
      '6xl': ['var(--font-size-6xl)', { lineHeight: '1' }],
    },
    extend: {
      colors,
      backgroundColor: { colors },
      borderColor: { colors },
      screens: {
        xs: '375px',
      },
    },
  },
  plugins: [
    plugin(({ addComponents, theme }) => {
      addComponents({
        '.focus-ui': {
          transitionProperty: theme('transitionProperty.colors'),
          transitionTimingFunction: theme('transitionTimingFunction.DEFAULT'),
          transitionDuration: theme('transitionDuration.500'),
        },
      });
    }),
  ],
};
